woocommerce/woocommerce-android
# Allow multiple products selections

Summary
==========
Fix issue #6046 by tracking the Product list inside the Order Creation view through the item ID instead of the Product ID, allowing the Order Creation feature to add the same product more than once.

It's important to notice that using the itemID to identify each added Product creates a constraint of having to sync that recently added Product before allowing the merchant to add another one since the `itemID` is created and defined by the Woo Core, this also introduces the constraint of having to protect the product quantity changes during the sync to avoid operating change requests on a Product where the `itemID` doesn't exist yet. To solve these issues, we apply the very same solution iOS does, by disabling Products interactions during the sync.

The main problem is that introducing a disable control to the quantity stepper view with the `ViewState` and loading lifecycle controls we have right now causes every single quantity change to disable everything immediately, causing a huge loss in the user experience when changing the product quantity, to solve that I decided to create a semantic definition of the loading start before the debounce delay ends inside the `CreateOrUpdateOrderDraft` use case, calling it `WillStart`. So when we change the `OrderDraft`, we immediately emit the `WillStart` Status update, after the debounce delay ends, the update effectively starts, emitting the `Ongoing` Status update. 

So now the Order Creation Fragment reacts to the update status changes in the following strategy:

- WillStart emitted -> Disable All price modifiers Add buttons
- Ongoing emitted -> Show loading progress, disable Product quantity stepper view **AND** keep all price modifiers Add - button disabled
- Succeeded or Failed emitted -> Enable everything back

@hichamboushaba since you have a lot of experience with the Order Creation draft update reactive structure, I would like to ask for your review of these changes if possible. Thanks! And since I'll be AFK for the next two weeks, there's no rush to review this one as this improvement is for platform parity with iOS and it's not critical.

Screenshots
==========
## Before
https://user-images.githubusercontent.com/5920403/161165568-0d270e72-4442-49ed-81a1-f17b1cadba80.mp4
## After
https://user-images.githubusercontent.com/5920403/161337863-98d468f1-5481-4c05-912c-97a02750dd38.mp4

How to Test
==========
1. Go to the Order Creation view
2. Select any product through the `Add Product` button
3. Verify that after adding the product, the `Add Product` button is disabled meanwhile the order is syncing
4. Hit the `Add Product` button once again and select the very same product from before
5. Verify that the product now is displayed twice, allowing distinct quantities from each other

Update release notes:
